### PR TITLE
Add methods for pipeline manipulation

### PR DESCRIPTION
**Pull request by mkosiarc**
CLOUDBLD-6301

Signed-off-by: mkosiarc <mkosiarc@redhat.com>

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] JSON/YAML configuration changes are updated in the relevant schema
- [ ] Changes to metadata also update the documentation for the metadata
- [ ] Pull request has a link to an osbs-docs PR for user documentation updates

### CODE REVIEW

File: osbs/tekton.py [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+            logger.error("[%s] 'Location' header is missing in response, cannot retrieve token",
+                         r.status_code)
+            return ""
+        parsed_url = urlparse(redir_url)
+        fragment = parsed_url.fragment
+        parsed_fragment = parse_qs(fragment)
+        self.token = parsed_fragment['access_token'][0]
+        return self.token
+
+    def get_serviceaccount_tokens(self, username="~"):
```
Comment (MartinBasti) [inline osbs/tekton.py:223]: Can you remind me why do we need this?

File: osbs/osbs_http.py [hunk 2]
```diff
@@ -54,6 +54,9 @@ def post(self, url, **kwargs):
     def put(self, url, **kwargs):
```
[the remark this answers is not in the document]
Comment (mkosiarc) [reply osbs/osbs_http.py:54]: not duplicated, renamed as there were some issues when importing stuff such as "from http import HTTPStatus"
"

File: osbs/tekton.py [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+                "namespace": self.namespace,
+                "labels": labels
+            },
+            "spec": {
+                "pipelineRef": {
+                    "name":self.pipeline_name
+                }
+            }
+        }
+        return self.os._patch(url, data=json.dumps(data),
```
Comment (MartinBasti) [inline osbs/tekton.py:432]: Will this update or rewrite all labels?

Original implementation updates labels, i.e. it first downloads current labels

File: osbs/tekton.py [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+            self.wait_for_pod()
+        url = self.os._build_url(
+            self.api_path,
+            self.api_version,
+            f"pods/{self.pod_name}"
+        )
+        r = self.os._get(url)
+        return r.json()
+
+    def get_logs(self, follow=False, wait=False):
```
Comment (MartinBasti) [inline osbs/tekton.py:584]: I'm very confused about return value and return type

File: osbs/tekton.py [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+        for changetype, pod in self.os.watch_resource(self.api_path, self.api_version, resource_type='pods', resource_name=self.pod_name):
+            try:
+                status = pod['status']['phase']
+            except KeyError:
+                logger.error("Pod does not have any status")
+                continue
+            if status in ['Running', 'Succeeded', 'Failed']:
+                return pod
+            else:
+                logger.debug("Pod is not in the state I am waiting for")
```
Comment (MartinBasti) [inline osbs/tekton.py:683]: Would be nice to mention what is the status

File: osbs/osbs_http.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkosiarc) [reply osbs/osbs_http.py:53]: Hmm, so there was still some issues in the dependencies. and anyway http.py module uses http.client.IncompleteRead exception, so I would prefer to just change the name

File: osbs/tekton.py (hunk 5 above)
Comment (mkosiarc) [reply osbs/tekton.py]: updated

File: osbs/tekton.py (hunk 1 above)
Comment (mkosiarc) [reply osbs/tekton.py:222]: I am not completely sure this going to stay the way it is - https://github.com/containerbuildsystem/osbs-client/blob/master/osbs/cli/main.py#L516

Comment (MartinBasti) [reply osbs/tekton.py:222]: aha, probably we should also do some cleanup of CLI commands (not here)

File: osbs/osbs_http.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MartinBasti) [reply osbs/osbs_http.py:53]: ok

File: osbs/tekton.py [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+                # Are we running inside a pod? If so, we will have a
+                # token available which can be used for authentication
+                self.use_auth = self.can_use_serviceaccount_token()
+        else:
+            self.use_auth = use_auth
+            if not auth_credentials_provided:
+                # We've been told to use authentication but no
+                # credentials have been given. See if we're running
+                # inside a pod, and if so use the provided token.
+                self.can_use_serviceaccount_token()
```
Comment (chmeliik) [inline osbs/tekton.py]: Is it ok to just discard the return value here? Shouldn't this raise an exception if `use_auth and not auth_credentials_provided and not self.can_use_serviceaccount_token()`?

**Review by chmeliik — COMMENTED** [review]
Ok, so I got part of the way through the PR before I realized I was reviewing copy-pasta'd code from osbs/core.py (man, I hate so much about it).

@mkosiarc could you help me and other reviewers understand what is new here? Ideally in commit messages?

Feel free to ignore the other comments, I'm just going to leave them here out of spite for the old code :smile:

File: osbs/tekton.py [hunk 7]
```diff
@@ -590,180 +355,357 @@ def log_and_sleep():
...
-            try:
-                build_response = self.wait(build_id, BUILD_FINISHED_STATES)
-                return build_response
-            except OsbsWatchBuildNotFound:
-                # this is woraround for https://github.com/openshift/origin/issues/2348
-                logger.warning("I'm going to wait again. Retry #%d.", retry)
-                continue
-        raise OsbsException("Failed to wait for a build: %s" % build_id)
+                # (Unknown, Started), (Unknown, PipelineRunCancelled)
+                logger.debug("Waiting for pipeline run, current status %s, reason %s", status, reason)
```
Comment (chmeliik) [inline osbs/tekton.py:484]: It's a bit confusing that Started doesn't mean started (at least according to this method). Could that be explained in a comment or docstring? Or maybe the method name needs to change?

Comment (chmeliik) [reply osbs/tekton.py:484]: Also, what should happen if the pipeline doesn't start? Can the last status be something other than 'True', 'False'? The method would just implicitly return None in that case

File: osbs/tekton.py [hunk 8]
```diff
@@ -590,180 +355,357 @@ def log_and_sleep():
...
+            except KeyError:
+                logger.error("pipeline run does not have any task runs")
                 continue
-        raise OsbsException('Failed to schedule a build in {} attempts: {}'.format(WAIT_RETRY,
-                                                                                   build_id))
+            # pipeline run finished succesfully or failed
+            if status in ['True', 'False']:
+                yield task_runs
+            elif status == 'Unknown' and reason == 'Running':
+                yield task_runs
```
Comment (chmeliik) [inline osbs/tekton.py:499]: I'm having a hard time understanding what the method does. It yields a list of task runs every time the status changes, that much I understand, but what does that list actually mean? Is it a list of task runs currently in progress? A list of all the task runs that were created up until that point? Something else?

Would it make sense to de-duplicate the taskruns here and flatten the result from `Iterator[List[str]]` to `Iterator[str]`?

And is it necessary to check the status+reason?

**Review by chmeliik — COMMENTED** [review]
Got a couple comments for PipelineRun, some of them also apply to TaskRun and maybe Pod.

Those three classes are similar in many ways, should they share a base class that implements the common stuff?

File: osbs/tekton.py (hunk 8 above)
Comment (lkolacek) [reply osbs/tekton.py:499]: +1 can you add explanation docstring to this method please?

File: osbs/tekton.py [hunk 9]
```diff
@@ -1,33 +1,20 @@
-"""
-Copyright (c) 2015, 2019 Red Hat, Inc
-All rights reserved.
-
-This software may be modified and distributed under the terms
-of the BSD license. See the LICENSE file for details.
-"""
-from __future__ import print_function, unicode_literals, absolute_import, division
 import json
```
[the remark this answers is not in the document]
Comment (mkosiarc) [reply osbs/tekton.py:7]: Fixed

File: osbs/tekton.py [hunk 10]
```diff
@@ -590,180 +355,357 @@ def log_and_sleep():
...
+class PipelineRun():
+    def __init__(self, os, pipeline_name, pipeline_run_name):
+        self.os = os
+        self.pipeline_name = pipeline_name
+        self.pipeline_run_name = pipeline_run_name
+        self.api_path = 'apis'
+        self.api_version = API_VERSION
+
+    def start_pipeline_run(self):
+        url = self.os._build_url(
```
[the remark this answers is not in the document]
Comment (mkosiarc) [reply osbs/tekton.py]: Made those methods public

File: osbs/tekton.py (hunk 8 above)
Comment (mkosiarc) [reply osbs/tekton.py]: I added explanation to docstring, and cleared up that method as per adam's suggestions.

File: osbs/tekton.py [hunk 11]
```diff
@@ -590,180 +355,357 @@ def log_and_sleep():
...
+            },
+            "spec":{
+                "pipelineRef":{
+                    "name":self.pipeline_name
+                },
+                "status": "PipelineRunCancelled"
+            }
+        }
+        return self.os._patch(url, data=json.dumps(data),
+                          headers={"Content-Type": "application/merge-patch+json", "Accept": "application/json"})
```
[the remark this answers is not in the document]
Comment (mkosiarc) [reply osbs/tekton.py]: Added docs to commit msg.

File: osbs/tekton.py (hunk 7 above)
Comment (mkosiarc) [reply osbs/tekton.py:442]: For your second comment - I think the last status cannot be anything other than True or False. So I am expecting that if the start pipeline code returns 200 OK status code, then the pipeline will have one of the statuses mentioned, Unknown, True or False.

Comment (chmeliik) [reply osbs/tekton.py:442]: Ack, the table here pretty much answered both my questions -> https://tekton.dev/docs/pipelines/pipelineruns/#monitoring-execution-status

I'm fine with keeping the name, maybe just include that link as an explanation

Comment (mkosiarc) [reply osbs/tekton.py]: kept the name, added doc link to method docstring

**Review by chmeliik — APPROVED** [review]
LGTM, assuming the duplication between tekton.py and core.py is removed later

**Pull request merged**